core_impl/src/test/java/com/google/instrumentation/stats/StatsContextTest.java

```diff
@@ -167,17 +169,21 @@ public void testToString() {
   }
 
   private static void testSerialize(Tag... tags) throws IOException {
+    final Map<TagKey, TagValue> tagMap = new HashMap<TagKey, TagValue>();
+    for (Tag tag : tags) {
+      tagMap.put(tag.getKey(), tag.getValue());
+    }
+
     ByteArrayOutputStream actual = new ByteArrayOutputStream();
+    new StatsContextImpl(tagMap).serialize(actual);
+
     ByteArrayOutputStream expected = new ByteArrayOutputStream();
     expected.write(VERSION_ID);
-    StatsContext.Builder builder = DEFAULT.builder();
-    for (Tag tag : tags) {
-      builder.set(tag.getKey(), tag.getValue());
+    for (Entry<TagKey, TagValue> entry : tagMap.entrySet()) {
       expected.write(VALUE_TYPE_STRING);
-      encodeString(tag.getKey().toString(), expected);
-      encodeString(tag.getValue().toString(), expected);
+      encodeString(entry.getKey().asString(), expected);
+      encodeString(entry.getValue().asString(), expected);
     }
-    builder.build().serialize(actual);
 
     assertThat(actual.toByteArray()).isEqualTo(expected.toByteArray());
```

Comment:
It looks like the order of the tags in `actual` depends on the implementation of the StatsContextImpl constructor.  The test could fail if the constructor copied the map.  A more reliable way to test the serialized format is to create one ByteArrayOutputStream for each possible order of the tags and then assert that the result matches one of them.